OEM: Provide Hetzner Images

apricote opened the pull request:
# Provide Hetzner OEM images

Adds build instructions for Hetzner Cloud images.

Pulls in https://github.com/flatcar/init/pull/118 and https://github.com/flatcar/bootengine/pull/94 in a temporary commit, will be removed once they are merged and used in `main`.

## How to use

The process to setup a server using these changes is described here: https://github.com/apricote/flatcar-packer-hcloud/blob/oem-image/README.md

## Testing done

I have built the image from this PR (based on an earlier nightly tag) and created a Hetzner Cloud server from it. I have tested the following (custom) functionality:

- Logging in with SSH Key
- Correct Hostname
- Afterburn Metadata File

For the usual Hetzner Cloud OS images, there is some "magic" happening through [hc-utils](https://github.com/hetznercloud/hc-utils) which is not included in this PR. I have not tested any functionality related to this.

Changelog Entry will be added closer to merge, as this needs a rebase anyway to get rid of the `tmp` commit.

- [ ] Changelog entries added in the respective `changelog/` directory (user-facing change, bug fix, security fix, update)
- [ ] Inspected CI output for image differences: `/boot` and `/usr` size, packages, list files for any missing binaries, kernel modules, config files, kernel modules, etc.

<!-- For coreos-overlay ebuild modifications that include a CROS_WORKON_COMMIT bump, did you bump too the ebuild revision ? -->

pothos commented on build_library/vm_image_util.sh:
```diff
@@ -278,6 +280,13 @@ IMG_azure_OEM_USE=azure
 IMG_azure_OEM_SYSEXT=oem-azure
 
+## hetzner
+IMG_hetzner_DISK_FORMAT=raw
```
```suggestion
```
Since this is the default I think we can omit it

apricote replied:
Removed in [c0d1796](https://github.com/flatcar/scripts/pull/1880/commits/c0d179620f1f0e6c51fbd439486064b52c82f982)

apricote commented on build_library/vm_image_util.sh:
```diff
@@ -278,6 +280,13 @@ IMG_azure_OEM_USE=azure
+IMG_hetzner_DISK_LAYOUT=vm
+IMG_hetzner_OEM_USE=hetzner
+IMG_hetzner_OEM_PACKAGE=common-oem-files
+# IMG_hetzner_OEM_SYSEXT=
```
Added in [c0d1796](https://github.com/flatcar/scripts/pull/1880/commits/c0d179620f1f0e6c51fbd439486064b52c82f982)

apricote commented on sdk_container/src/third_party/coreos-overlay/coreos-base/afterburn/files/coreos-metadata.service:
```diff
@@ -19,6 +19,9 @@ ConditionKernelCommandLine=|coreos.oem.id=packet
 ConditionKernelCommandLine=|flatcar.oem.id=scaleway
 
+ConditionKernelCommandLine=|flatcar.oem.id=hetzner
+ConditionKernelCommandLine=|coreos.oem.id=hetzner
```
Removed in [9a4705a](https://github.com/flatcar/scripts/pull/1880/commits/9a4705a6d9d3b6e2f43cc6ba10e8aa4eeafde4d0)

apricote commented:
@tormath1 [built a preview image](https://matrix.to/#/!SakMJmpDOgVMlGukUp:matrix.org/$KLwsUDLH4uOlUEcbC-oo2MPDqLECVF2RdoCmTgU-ok8?via=matrix.org&via=seymour.family&via=acrawford.com) for this PR (9a4705a). It is available here for download: https://bincache.flatcar-linux.net/images/amd64/9999.0.0+hetzner/

I updated my Packer documentation to use this image instead of using a locally built one. New docs + Packer templates are available on a new branch: https://github.com/apricote/flatcar-packer-hcloud/tree/oem-image-prebuilt

apricote commented on sdk_container/src/third_party/coreos-overlay/coreos-base/coreos-init/coreos-init-9999.ebuild:
Is this something that should be done in this PR? My plan was to remove 69a51ebf61599f1011918c70de7c053b9d64be67 from this PR once `init` & `bootengine` PRs are merged and the new commits are referenced in `flatcar-master`.

apricote replied:
Fine by me! Will update the references, increment the revision & rebase now.

apricote replied:
Moved in 1cf9d7f1592ac436752b983e2bb7da3c06f08c59

apricote commented on sdk_container/src/third_party/coreos-overlay/coreos-base/common-oem-files/common-oem-files-0-r6.ebuild:
Done in e2e3b33692c0826da8976371dde9638d53980fe6

apricote commented on sdk_container/src/third_party/coreos-overlay/sys-kernel/bootengine/bootengine-9999.ebuild:
Moved in e2f6180b47be290b308b1aa49826cb175ac52ec9

apricote commented:
- Updated the commit references for `bootengine` and `coreos-init` to their current `flatcar-master` commits, which include the required changes for this PR. Also incremented their revisions by 1 each
- Rebased against `main` and squashed the commits
